flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractStreamingCommitterOperator.java

```diff
@@ -0,0 +1,177 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.operators.sink;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeutils.base.array.BytePrimitiveArraySerializer;
+import org.apache.flink.api.connector.sink.Committer;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.operators.util.SimpleVersionedListState;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+
+/**
+ * Abstract base class for operators that work with a {@link Committer} or a {@link org.apache.flink.api.connector.sink.GlobalCommitter}.
+ *
+ * <p>Sub-classes are responsible for implementing {@link #recoveredCommittables(List)} and {@link #preCommit()}.
+ *
+ * @param <InputT> The input type of the {@link Committer}.
+ * @param <CommT> The committable type of the {@link Committer}.
+ */
+abstract class AbstractStreamingCommitterOperator<InputT, CommT> extends AbstractStreamOperator<CommT>
+		implements OneInputStreamOperator<InputT, CommT> {
+
+	private static final long serialVersionUID = 1L;
+
+	/** The operator's state descriptor. */
+	static final ListStateDescriptor<byte[]> STREAMING_COMMITTER_RAW_STATES_DESC =
+			new ListStateDescriptor<>(
+					"streaming_committer_raw_states",
+					BytePrimitiveArraySerializer.INSTANCE);
+
+	/** Group the committable by the checkpoint id. */
+	private final NavigableMap<Long, List<CommT>> committablesPerCheckpoint;
+
+	/** The committable's serializer. */
+	private final StreamingCommitterStateSerializer<CommT> streamingCommitterStateSerializer;
+
+	/** Responsible for committing the committable to the external system. **/
+	protected final Committer<CommT> committer;
+
+	/** The operator's state. */
+	private ListState<StreamingCommitterState<CommT>> streamingCommitterState;
+
+	/** Inputs collected between every pre-commit. */
+	protected List<InputT> currentInputs;
```

Comment:
The relationship between `currentInput` and `currentGlobalCommittables` seems to be a little complex to understand. As a whole, I would tend to 

1. Make `currentInputs` only available in the abstract class, pass it to `precommit` as a parameter and clear it in the abstract class.
2. In `StreamingCommitterOperator` the result of `recoveredCommittables` should be saved in a standalone list, as did in the `GlobalStreamingCommitterOperator`. The recovered committables are semantically not inputs.
3. In `StreamingCommitterOperator` and `GlobalStreamingCommitterOperator` rename the `currentGlobalCommittables` to be `recovered(Global)Committables`.  


One pending problem here is that do we need to immediately commit the recovered committables ? In the current implementation of `StreamingFileSink` we did so.